contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1487 @@
+        }
+
+        (uint256 keepPenalty, uint96 tRemainder) = tToKeep(tPenalty);
+        if (keepPenalty == 0) {
+            return tAmountToSlash;
+        }
+        tPenalty -= tRemainder;
+        operatorStruct.keepInTStake -= tPenalty;
+        tAmountToSlash -= tPenalty;
+
+        address[] memory operatorWrapper = new address[](1);
+        operatorWrapper[0] = operator;
+        keepStakingContract.seize(
+            keepPenalty,
+            rewardMultiplier,
+            msg.sender,
+            operatorWrapper
+        );
+        return tAmountToSlash;
+    }
+
+    /// @notice Convert amount from T to NU and call `slashStaker` in NuCypher staking contract.
+    ///         Returns remainder of slashing amount in T
+    function seizeNu(
+        OperatorInfo storage operatorStruct,
+        uint96 tAmountToSlash,
+        uint256 rewardMultiplier
+    ) internal returns (uint96) {
+        if (operatorStruct.nuInTStake == 0) {
+            return tAmountToSlash;
+        }
+
+        uint96 tPenalty;
+        if (tAmountToSlash <= operatorStruct.nuInTStake) {
+            tPenalty = tAmountToSlash;
+        } else {
+            tPenalty = operatorStruct.nuInTStake;
+        }
+
+        (uint256 nuPenalty, uint96 tRemainder) = tToNu(tPenalty);
+        if (nuPenalty == 0) {
+            return tAmountToSlash;
+        }
+        tPenalty -= tRemainder;
+        operatorStruct.nuInTStake -= tPenalty;
+        tAmountToSlash -= tPenalty;
+
+        uint256 nuReward = nuPenalty.percent(SLASHING_REWARD_PERCENT).percent(
+            rewardMultiplier
+        );
+        nucypherStakingContract.slashStaker(
+            operatorStruct.owner,
+            nuPenalty,
+            msg.sender,
+            nuReward
+        );
+        return tAmountToSlash;
+    }
+
+    /// @notice Removes application with zero authorization from authorized
+    ///         applications array
+    function cleanAuthorizedApplications(
+        OperatorInfo storage operatorStruct,
+        uint256 numberToDelete
+    ) internal {
+        uint256 length = operatorStruct.authorizedApplications.length;
+        if (numberToDelete == length) {
+            delete operatorStruct.authorizedApplications;
+            return;
+        }
+
+        uint256 deleted = 0;
+        uint256 index = 0;
+        uint256 newLength = length - numberToDelete;
+        while (index < newLength && deleted < numberToDelete) {
+            address application = operatorStruct.authorizedApplications[index];
+            if (operatorStruct.authorizations[application].authorized == 0) {
+                operatorStruct.authorizedApplications[index] = operatorStruct
+                    .authorizedApplications[length - deleted - 1];
+                deleted++;
+            } else {
+                index++;
+            }
+        }
+        for (index = newLength; index < length; index++) {
+            operatorStruct.authorizedApplications.pop();
+        }
```

Comment:
We could expose this function as public in a stub and implement more unit tests for various cases. It looks fine to me but better be safe than sorry.